### PR TITLE
feat: s3 backup store can save metadata

### LINKED ISSUE #10085: Backup store can save backup metadata
Create a 'metadata' object, prefixed the same as other objects, eg. `checkpointId/partitionId/nodeId/metadata`. The object should be human-readable, for example JSON, and contain all information to reconstruct the backup.

### PR DESCRIPTION
**Pull request by lenaschoenburg**
## Description

This sets up the s3 backup store module and implements the saving of backup metadata as a JSON file.

## Related issues

closes #10085

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    this.config = config;
+    this.client = client;
+  }
+
+  @Override
+  public CompletableFuture<Void> save(final Backup backup) {
+    final var metadata = saveMetadata(backup);
+
+    return CompletableFuture.allOf(metadata);
+  }
```
Comment (deepthidevaki) [inline backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:57]: :question:  Should this run on a separate thread to not block the caller? Or is it enough that client is async which uses its own executor?

The caller is BackupManager which is running on its own actor. I think it is better to not block it so that it can do other operations such as  responding to status query while taking a backup.

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                  .contentType("application/json")
+                  .build(),
+          AsyncRequestBody.fromBytes(MAPPER.writeValueAsBytes(Metadata.of(backup))));
+    } catch (JsonProcessingException e) {
+      return CompletableFuture.failedFuture(e);
+    }
+  }
+
+  private static String objectPrefix(BackupIdentifier id) {
+    return "%s/%s/%s/".formatted(id.checkpointId(), id.partitionId(), id.nodeId());
```
Comment (deepthidevaki) [inline backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:95]: :wrench: If I remember correctly, we decided to put in this order partitionId/checkpointId/nodeid

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/Metadata.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Set;
+
+/**
+ * Holds all metadata information about a backup, for example the {@link
+ * io.camunda.zeebe.backup.api.BackupIdentifier}, the {@link
+ * io.camunda.zeebe.backup.api.BackupDescriptor} and file names of segments and snapshot.
+ *
+ * <p>Used for JSON serialization.
+ */
+record Metadata(
```
Comment (deepthidevaki) [inline backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/Metadata.java:20]: :x: Add @JsonIgnoreProperties(ignoreUnknown = true) for backward compatibility when adding new fields.

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3Config.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.backup.s3;
+
+public record S3Config(String bucketName) {}
```
Comment (deepthidevaki) [inline backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3Config.java:10]: :wrench: Add documentation. Is this config related to backup only? Then I suggest it to rename accordingly - may be `S3BackupConfig`.

**Review by deepthidevaki — APPROVED** [review]
:rocket: Awesome. Happy to see the progress!

Please take a look at the comments before merging.

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java (hunk 2 above)
Comment (oleschoenburg) [reply backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:95]: Oops, good catch :+1:

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/Metadata.java (hunk 3 above)
Comment (oleschoenburg) [reply backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/Metadata.java:22]: I think that's forwards compatability, right? Old brokers can read metadata from newer brokers where new fields are added. Anyway, added the annotation :+1:

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3Config.java (hunk 4 above)
Comment (oleschoenburg) [reply backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3Config.java]: :+1: Adjusted the name, but I kept the documentation very light for now.

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java (hunk 1 above)
Comment (oleschoenburg) [reply backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:57]: This is something I haven't thought about much until now. The S3 async client uses it's own thread pool for IO but eventually the backup store will also have to wait for the upload to complete update the status. This is where we will have to be careful to use the same thread pool and not the callers thread. 

I think for now we can keep this as is, returning a CompletableFuture that is completed by the s3 client when uploading is finished is exactly what we want to do here. BackupManager can decide if it wants to wait for completion or not.
